algebird-core/src/main/scala/com/twitter/algebird/ExpHist.scala

```diff
@@ -0,0 +1,342 @@
+    }
+
+  // This internal method assumes that the instance is stepped forward
+  // already, and does NOT try to step internally. It also assumes
+  // that `items` is sorted in ASCENDING order, with newer items on
+  // the right side, and that `items.map(_.size).sum == delta`.
+  private[ExpHist] def addAllWithoutStep(items: Vector[Bucket], delta: Long): ExpHist = {
+    val inputs = items ++ buckets
+    val desiredBuckets = Canonical.bucketsFromLong(total + delta, conf.l)
+    copy(
+      buckets = ExpHist.rebucket(inputs, desiredBuckets),
+      total = total + delta)
+  }
+
+  def oldestBucketSize: Long = if (total == 0) 0L else buckets.last.size
+
+  /**
+   * Smallest possible count seen in the last conf.windowSize
+   * timestamps.
+   */
+  def lowerBoundSum: Long = total - oldestBucketSize
+
+  /**
+   * Largest possible count seen in the last conf.windowSize
+   * timestamps.
+   */
+  def upperBoundSum: Long = total
+
+  /**
+   * Estimate of the count seen across the last conf.windowSize
+   * timestamps. Guaranteed to be within conf.epsilon of the true
+   * count.
+   */
+  def guess: Double =
+    if (total == 0) 0.0
+    else (total - (oldestBucketSize - 1) / 2.0)
+
+  /**
+   * relative error of guess, guaranteed to be <= conf.epsilon.
+   */
+  def relativeError: Double =
+    if (total == 0) 0.0
+    else {
+      val minInsideWindow = total + 1 - oldestBucketSize
+      val absoluteError = (oldestBucketSize - 1) / 2.0
+      absoluteError / minInsideWindow
+    }
+}
+
+/**
+ * The paper that introduces the exponential histogram proves that,
+ * given a positive number `l`, every integer can be uniquely
+ * represented as the sum of
+ *
+ * (`l` or `l + 1`) * 2^i
+ *
+ * for i = 0 to j, given some j.
+ *
+ * The paper calls this the "l-canonical" representation of the
+ * number.
+ *
+ * It turns out that if you follow the exponential histogram
+ * bucket-merging algorithm, you end up with the invariant that the
+ * number of buckets with size 2^i exactly matches that power of 2's
+ * coefficient in its l-canonical representation.
+ *
+ * Put another way - only sequences of buckets with sizes matching
+ * the l-canonical representation are valid exponential histograms.
+ *
+ * (We use this idea in `ExpHist.rebucket` to take a sequence of
+ * buckets of any size and rebucket them into a sequence where the
+ * above invariant holds.)
+ *
+ * This is huge. This means that you can implement `addAll(newBuckets)` by
+ *
+ * - calculating newTotal = total + delta contributed by newBuckets
+ * - generating the l-canonical sequence of bucket sizes for newTotal
+ * - rebucketing newBuckets ++ oldBuckets into those bucket sizes
+ *
+ * The resulting sequence of buckets is a valid exponential
+ * histogram.
+ */
+object Canonical {
+  @inline private[this] def floorPowerOfTwo(x: Long): Int =
+    JLong.numberOfTrailingZeros(JLong.highestOneBit(x))
+
+  @inline private[this] def modPow2(i: Int, exp2: Int): Int = i & ((1 << exp2) - 1)
```

Comment:
what about the bugs on inlining you found? Are we sure none of these exhibit this? Downstream folks still could if they compile with `-optimize`.